### PR TITLE
findChessboardCorners performance degradation fix for some cases

### LINKED ISSUE #23558: cv2.findChessboardCorners() running endlessly on certain images
### System Information

OpenCV python version: both 3.4.18.65 and 4.7.0.72
Operating System / Platform: Ubuntu 18.04 and macOS 13.2.1
Python version: 3.7

### Detailed description

I’m using OpenCV to find chessboard corners on an image using `cv2.findChessboardCorners()`. It can happen that an image is passed into the function that is of bad quality or almost completely black. When this happens, the function just endlessly runs, trying to find the corners.

I initially opened a [topic on the OpenCV forum](https://forum.opencv.org/t/cv2-findchessboardcorners-running-endlessly-on-certain-images/12972), but there it was confirmed that it's a bug, hence why I'm also opening a bug report here.

This issue is a problem to me because it completely blocks my application when this happens.

### Steps to reproduce

This issue can easily be reproduced.

You can use the code snippet below:
```
import cv2

test = cv2.imread("/path/to/the/image/27-04-2023_14-13-21_494.png")

found, corners = cv2.findChessboardCorners(image=test,
                                           patternSize=(26, 16),
                                           flags=(cv2.CALIB_CB_FAST_CHECK +
                                                  cv2.CALIB_CB_ADAPTIVE_THRESH +
                                                  cv2.CALIB_CB_NORMALIZE_IMAGE))

print(found)
```

in combination with the image below:

You will notice that the function just endlessly keeps running and consuming resources (I’ve let it run for +15 minutes already).

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by thewoz**
Fixes #23558

In the issue #23558 an empty image was passed to the findChessboardCorners() function with the CALIB_CB_FAST_CHECK flag. The fast check failed and findChessboardCorners() wasted 15minutes finding a checkerboard that was not there in the image.

The reason is that the checkChessboardBinary() function returned true because the checkQuads() function was convinced that among all the contours found by fillQuads() were those of the checkerboard.

The problem was icvGetQuadrangleHypotheses() that incorrectly filtering the contours found.  It was passing too many contours to checkQuads(), and by the law of large numbers, checkQuads() was able to find contours that "looked" like a checkerboard between them.

What I did is to shrink the parameters in the icvGetQuadrangleHypotheses() function.  I changed the range of the ratio between the sides of the contour found to 0.5 and 1.5. More reasonable values than the previous 0.3 and 3.0.  In addition, I raised the minimum side length of the hypothetical square from 10 to 25 pixels.

After the change I tried the function on other checkerboard images and had no problems.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@thewoz Thanks a lot for the contribution. The PR introduces existing test regression:
```
[ RUN      ] Calib3d_ChessboardDetector.timing
C:\build\precommit_windows64\4.x\opencv\modules\ts\src\ts.cpp(618): error: Failed

	failure reason: Invalid function output
	test case #-1
	seed: 00000000000c5a7c
-----------------------------------
	LOG:
chess3.png: chessboard 1:
Error: chessboard was not detected in the image chess3.png

-----------------------------------

[  FAILED  ] Calib3d_ChessboardDetector.timing (7 ms)
```

**Comment by thewoz** [comment]
sorry where I can found the chess3.png images?

**Comment by s-trinh** [comment]
@thewoz 
It should be this image: https://github.com/opencv/opencv_extra/blob/38d14db05e43edd9f44e673cfd296ff67a210222/testdata/cv/cameracalibration/chess3.png

**Comment by thewoz** [comment]
Hi,
I have test the pull request against the images:
BoardStereoL1-6.png BoardStereoR1-6.png and chess1-6.png 

in 

https://github.com/opencv/opencv_extra/tree/4.x/testdata/cv/cameracalibration

and the chessboard was found in each of them.
so is not clear to me the failed in the test

**Comment by vpisarev** [comment]
@thewoz, basically, you should clone opencv_extra repository and then set OPENCV_TEST_DATA_PATH environment variable pointing to <opencv_extra_dir>/testdata.

**Comment by thewoz** [comment]
Hi,
I should have fixed the problem on the image chess3.png

**Comment by AleksandrPanov** [comment]
```
[ RUN      ] Calib3d_ChessboardDetector.timing
/home/ci/opencv/modules/ts/src/ts.cpp:618: Failure
Failed

	failure reason: Invalid function output
	test case #4
	seed: 00000000000c5a60
-----------------------------------
	LOG:
chess3.png: chessboard 1:
    cvCheckChessboard time s: 0.000000, us per pixel: 0.000000
    cvFindChessboard time s: 0.000000, us per pixel: 0.000000
chess1.png: chessboard 1:
    cvCheckChessboard time s: 0.000000, us per pixel: 0.000000
    cvFindChessboard time s: 0.000000, us per pixel: 0.000000
../shared/baboon.png: chessboard 0:
    cvCheckChessboard time s: 0.000000, us per pixel: 0.000000
    cvFindChessboard time s: 0.000000, us per pixel: 0.000000
../shared/box_in_scene.png: chessboard 0:
    cvCheckChessboard time s: 0.000000, us per pixel: 0.000000
    cvFindChessboard time s: 0.000000, us per pixel: 0.000000
../shared/airplane.png: chessboard 0:
    cvCheckChessboard time s: 0.000000, us per pixel: 0.000000
    cvFindChessboard time s: 0.000000, us per pixel: 0.000000
chessboard-artificial2.png: chessboard 1:
Error: chessboard was not detected in the image chessboard-artificial2.png

-----------------------------------
	CONSOLE: ...
-----------------------------------

[  FAILED  ] Calib3d_ChessboardDetector.timing (164 ms)
```

**Comment by thewoz** [comment]
Hello everyone,
everyone seems to be working now.
All tests are passed

**Comment by thewoz** [comment]
sorry I forgot to erase a include file that I use for the debug

**Comment by thewoz** [comment]
Hello everyone is there anything else I can do ?

**Comment by asmorkalov** [comment]
@vpisarev could you take a look too?

**Comment by thewoz** [comment]
is there anything i can do?

**Review by vpisarev — APPROVED** [review]

**Comment by vpisarev** [comment]
@thewoz, thank you! I think, the patch can be merged.

File: modules/calib3d/src/checkchessboard.cpp [hunk 1]
```diff
@@ -75,7 +76,14 @@ static void icvGetQuadrangleHypotheses(const std::vector<std::vector< cv::Point
...
         }
 
+        float box_area = box.size.width * box.size.height;
+        if(contourArea(c) < (box_area_covered * box_area))
+        {
+          continue;
+        }
+
         quads.emplace_back(box_size, class_id);
+
```
Comment (AleksandrPanov) [inline modules/calib3d/src/checkchessboard.cpp:86]: extra `\n`

File: modules/calib3d/src/checkchessboard.cpp [hunk 2]
```diff
@@ -96,6 +104,7 @@ inline bool less_pred(const std::pair<float, int>& p1, const std::pair<float, in
 static void fillQuads(Mat & white, Mat & black, double white_thresh, double black_thresh, vector<pair<float, int> > & quads)
 {
     Mat thresh;
+
```
Comment (AleksandrPanov) [inline modules/calib3d/src/checkchessboard.cpp:107]: extra `\n`

File: modules/calib3d/src/checkchessboard.cpp [hunk 3]
```diff
@@ -111,6 +120,7 @@ static void fillQuads(Mat & white, Mat & black, double white_thresh, double blac
         findContours(thresh, contours, hierarchy, RETR_CCOMP, CHAIN_APPROX_SIMPLE);
         icvGetQuadrangleHypotheses(contours, hierarchy, quads, 0);
     }
+
```
Comment (AleksandrPanov) [inline modules/calib3d/src/checkchessboard.cpp:123]: extra `\n`

**Review by AleksandrPanov — CHANGES_REQUESTED** [review]
1. need to fix extra `\n`
2. what is the performance? reabase to https://github.com/opencv/opencv/pull/24605#event-11212616852, this PR improves performance.
3. I suggest to use `approxPolyDP()` instead `box_area_covered`

File: modules/calib3d/src/checkchessboard.cpp [hunk 4]
```diff
@@ -74,8 +77,15 @@ static void icvGetQuadrangleHypotheses(const std::vector<std::vector< cv::Point
...
             continue;
         }
-
+      
+        float box_area = box.size.width * box.size.height;
+        if(contourArea(c) < (box_area_covered * box_area))
+        {
+          continue;
+        }
+
```
[the remark this answers is not in the document]
Comment (thewoz) [reply modules/calib3d/src/checkchessboard.cpp]: Hi @AleksandrPanov I never perform a rebase.
Can you help me? Sorry about that...

[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/calib3d/src/checkchessboard.cpp]: checkout your branch to 4.x, update your 4.x branch from https://github.com/opencv/opencv, checkout your branch to [fix-issue-23558](https://github.com/thewoz/opencv/tree/fix-issue-23558), rebase and resolve conflicts:
```
git checkout 4.x
git pull https://github.com/opencv/opencv
git checkout fix-issue-23558
git rebase 4.x
git push -f
```

**Comment by thewoz** [comment]
Hello everyone,
I was trying the modification suggested by @AleksandrPanov to try to use the approxPolyDP() function and it wasn't working too well.
Anyway doing some testing I realized that today's version of 4.x no longer gives the issue #23558 
I think that the pull [fix-issue-23558](https://github.com/thewoz/opencv/tree/fix-issue-23558), has solved this #23558 
At this point I would say that this pull of mine is useless.

**Comment by AleksandrPanov** [comment]
@thewoz, I tried to use the approxPolyDP() func too, and it really wasn't working too well. You could try to call this function twice, but it doesn't look like a good solution.

Now I think that use `minAreaRect` and `box_area_covered` is better. Thanks for the good idea.

**Comment by thewoz** [comment]
Hi @AleksandrPanov... yes, we could try to run approxPolyDP() twice (I haven't tried yet). 
However, my concern is that approxPolyDP() is a bit too sensitive to the sharpeness of the edges of the checkerboard squares.
My idea is that using `minAreaRect` and `box_area_covered` would make the process a little less sensitive to this problematic.
Anyway, the reason I tried to implement this solution is to solve problem https://github.com/opencv/opencv/issues/23558, in which too many contours were found in an image and `checkChessboardBinary()` thought there was a checkerboard. 
So I wanted to try to implement a method that would remove those contours that could not be considered square. 
Up to now the pull request #24605 seems to have solved the problem https://github.com/opencv/opencv/issues/23558

**Comment by AleksandrPanov** [comment]
@thewoz, there are not enough tests in OpenCV for findchessboardcorners() and therefore algorithm changes can cause regressions that we may not catch.
We merged PR #24605 because it just adds optimizations without changing the algorithm.

We want to expand the test suite, but this is a big task. After that, we will try to improve the code based on your ideas.

**Comment by thewoz** [comment]
Hi @AleksandrPanov,
I completely understand what you mean and I fully agree with you.
Thank you for all the work

**Pull request closed without merging**